### PR TITLE
Convolution transpose

### PR DESCRIPTION
**Pull request by hunse**
**Motivation and context:**

Transpose convolution is a fairly general way to do different types of upsampling in deep networks. Adding this transform allows many more types of deep networks to be implemented in Nengo, including many types of object detection networks.

**How has this been tested?**

I added a test that ensures that the `ConvTransposeInc` op is the transpose of `ConvInc`. Since `ConvInc` is tested against the Scipy implementation, if all tests pass then `ConvTransposeInc` (and the underlying code we've borrowed from `np-conv2d`) should be working fine. I've also added tests for the API (mostly piggy-backing off the existing `Convolution` test).

**How long should this take to review?**

- Average (neither quick nor lengthy)

**Where should a reviewer start?**

Start with the `ConvolutionTranspose` transform, since this is the API (it's fairly similar to the `Convolution` transform). Then, the new test in `builder/tests/test_transforms.py` is the main test that's added to test the underlying implementation. The code in `_vendors/np-conv2d` is copied from [np-conv2d](https://github.com/renmengye/np-conv2d), with the modifications listed in the `conv2d_gradx` docstring.

**Types of changes:**

- New feature (non-breaking change which adds functionality)

**Checklist:**

- [x] I have read the **CONTRIBUTING.rst** document.
- [x] I have updated the documentation accordingly.
- [x] I have included a changelog entry.
- [x] I have added tests to cover my changes.
- [x] I have run the test suite locally and all tests passed.

**Still to do:**

- [x] Changelog entries
- [x] Check that it's showing in the API docs properly

### CODE REVIEW

File: nengo/builder/tests/test_transforms.py [hunk 1]
```diff
@@ -109,9 +109,169 @@ def test_convinc_2d(
...
+        step_trans()
+        AtCt.append(signals_tr[yt].copy().ravel())
+
+    AtCt = np.array(AtCt)
+    conv_transpose_output = AtCt.dot(x_flat)
+    assert conv_transpose_output.shape == conv_output.shape
+
+    # Check whether the outputs match. If they don't, print/plot some diagnostic output.
+    success = allclose(conv_transpose_output, conv_output)
+    if not success:
```
Comment (drasmuss) [inline nengo/builder/tests/test_transforms.py:194]: This seems like a lot of code just to diagnose test failures (which seems like it'd be of pretty limited utility after this PR is merged). If we don't want to lose this code snippet I'd be inclined to just add it as a comment on the PR (and then just `assert success` here).

File: nengo/builder/transforms.py [hunk 2]
```diff
@@ -6,7 +6,13 @@
 from nengo.builder.signal import Signal
 from nengo.exceptions import BuildError
 from nengo.rc import rc
-from nengo.transforms import Convolution, Dense, NoTransform, Sparse
+from nengo.transforms import (
+    Convolution,
+    ConvolutionTranspose,
```
Comment (drasmuss) [inline nengo/builder/transforms.py:11]: Seeing this made me think, maybe we should add `Conv`/`ConvTranspose` aliases for these?

File: nengo/builder/transforms.py [hunk 3]
```diff
@@ -158,6 +170,10 @@ def __init__(self, W, X, Y, conv, tag=None):
         self.reads = [W, X]
         self.updates = []
 
+    @property
+    def transpose(self):
```
Comment (drasmuss) [inline nengo/builder/transforms.py:174]: I'd use `is_transpose` here (just so that there's no confusion with something like `ndarray.transpose`, which actually transforms the object to be a transposed version).

File: nengo/transforms.py [hunk 4]
```diff
@@ -508,6 +451,219 @@ def output_shape(self):
...
+
+class ConvolutionTranspose(_ConvolutionBase):
+    """An N-dimensional transposed convolutional transform.
+
+    This performs the transpose operation of `.Convolution`. The ``kernel_size``,
+    ``strides``, and ``padding`` parameters all act as in `.Convolution`, so this
+    transform will be the transpose of a `.Convolution` transform with those
+    parameters. The ``n_filters`` and ``input_shape`` parameters are relative to *this*
+    transform. The output shape is ambiguous, and can thus be specified (i.e. with
+    `.Convolution`, there can be more than input shape that produces the same output
```
Comment (drasmuss) [inline nengo/transforms.py:544]: typo -> "more than one"

**Review by drasmuss — COMMENTED** [review]
Looks good, just a couple minor comments/questions.

File: nengo/_vendor/npconv2d/conv2d.py [hunk 5]
```diff
@@ -112,6 +107,67 @@ def extract_sliding_windows(x, ksize, pad, stride, floor_first=True):
...
+        x, ((0, 0), pph, ppw, (0, 0)),
+        mode='constant',
+        constant_values=(0.0, ))
+
+    # The following code extracts window without copying the data:
+    # y = np.zeros([n, h2, w2, kh, kw, c])
+    # for ii in range(h2):
+    #     for jj in range(w2):
+    #         y[:, ii, jj, :, :, :] = x[:, ii:ii + kh, jj:jj + kw, :]
+    x_sn, x_sh, x_sw, x_sc = x.strides
```
[the remark this answers is not in the document]
Comment (hunse) [reply nengo/_vendor/npconv2d/conv2d.py:174]: Both! The commented-out code was the original, and I made a PR there to change it to the active code. When he merged it, he kept the old implementation commented out. https://github.com/renmengye/np-conv2d/pull/1/files

[the remark this answers is not in the document]
Comment (hunse) [reply nengo/_vendor/npconv2d/conv2d.py:174]: So what's here is identical to `np-conv2d:master`

File: nengo/_vendor/npconv2d/conv2d.py [hunk 6]
```diff
@@ -131,3 +187,48 @@ def conv2d(x, w, pad='SAME', stride=(1, 1)):
...
+        dy: [N, H', W', K]
+        w: [I, J, K, C]
+        xsize: Original image size, [H, W]
+
+    Returns:
+        dx: [N, H, W, C]
+
+    Modifications from original code:
+      - Do not transpose input/output channels in `w`
+      - Fix bug in computing `pad2w` (use `dys[1]` instead of `dys[0]`)
```
[the remark this answers is not in the document]
Comment (hunse) [reply nengo/_vendor/npconv2d/conv2d.py:218]: Sure, I can. I've got a PR to fix it there, and it's the only PR, so I feel like if there are any updates to that repo, they'll include this fix. renmengye/np-conv2d#2

[the remark this answers is not in the document]
Comment (drasmuss) [reply nengo/_vendor/npconv2d/conv2d.py:218]: Ah if you already have an upstream PR then that's probably fine (could link it here I guess?)

File: nengo/builder/tests/test_transforms.py (hunk 1 above)
Comment (hunse) [reply nengo/builder/tests/test_transforms.py:194]: If we're having it as a comment here, I don't see the benefit of that versus having it be active code. It's something that I think I tried to get rid of more than once, and then found another bug and realized I needed it again. It can be a real pain to diagnose these bugs, and there's no guarantees more won't come up in the future (there are so many parameters, and thus so many edge cases).

File: nengo/builder/transforms.py (hunk 2 above)
Comment (hunse) [reply nengo/builder/transforms.py:11]: It would certainly look more familiar for those coming from many other places, like TensorFlow/Keras. I would be fine with adding those.

File: nengo/builder/tests/test_transforms.py (hunk 1 above)
Comment (drasmuss) [reply nengo/builder/tests/test_transforms.py:194]: It just seems odd to me that almost half the code in this test is not actually part of the test, it's just there to help with debugging in case the test fails. That's certainly not something we've typically done throughout the code base.

**Comment by drasmuss** [comment]
Fixups added. I was thinking I'd squash all the convolution-related commits (so everything except the node init ordering and test_seeding), unless you think there are any that are important to keep separate?

**Comment by hunse** [comment]
The fixups look good. Squashing all the convolution ones is fine by me.

**Comment by hunse** [comment]
@drasmuss Got this rebased onto master, and made the appropriate f-string changes (I didn't see any OrderedDicts or places for pathlib). I don't think all the fixups necessarily point to the right base commit, but if you're squashing together all the convolution stuff anyway it should be fine.

**Comment by hunse** [comment]
I've fixed the alias problem in nengo/nengo-sphinx-theme#70. This now requires that.

**Comment by hunse** [comment]
This is ready for another look @drasmuss, when you get a chance.

**Review by tbekolay — CHANGES_REQUESTED** [review]
Mostly LGTM, but this PR adds a lot of noise to the test suite that we can hopefully pare down before merging. Also, we've brought in TensorFlow as an optional test dependency, which isn't ideal, because we don't want to really want to mark it as an optional dependency because it doesn't do anything except for run tests, but we also don't want to mark it as a test dependency because it's a very big dependency that tests a small fraction of the codebase. Is there a simple-ish way to implement those tests without TensorFlow? If not, I think we will probably want to figure out some way to ensure that these tests run at least as part of the release process, as otherwise they won't get run regularly.

File: nengo/tests/test_transforms.py [hunk 7]
```diff
@@ -6,6 +6,44 @@
 from nengo.exceptions import BuildError, ValidationError
 
 
+@pytest.mark.parametrize("x_mul", (1, 2, 3, 4))
+@pytest.mark.parametrize("k_size", (1, 2, 3, 4))
+@pytest.mark.parametrize("stride", (1, 2, 3, 4))
+@pytest.mark.parametrize("padding", ("same", "valid"))
+def test_convolution_shape(padding, stride, k_size, x_mul, rng, allclose):
```
[the remark this answers is not in the document]
Comment (hunse) [reply nengo/tests/test_transforms.py:13]: It's tricky, because all of them cover potential edge cases.

Probably the best way to reduce the number of tests is rather than covering all combinations of all cases, pick a smattering that together cover all the different values of each parameter (basically, do a random search instead of a grid search).

If we want to be thorough, we could even have the full test (all combinations) run as part of our "slow" tests that we run on release (though honestly, it's actually quite fast to go through even all 128 combinations).

File: nengo/builder/tests/test_transforms.py [hunk 8]
```diff
@@ -109,9 +110,213 @@ def test_convinc_2d(
     assert allclose(signals[y], y0)
 
 
+@pytest.mark.parametrize("channels_last", (True, False))
+@pytest.mark.parametrize("strides", [(1, 1), (2, 2), (3, 2)])
+@pytest.mark.parametrize("kernel_size", [(4, 5), (5, 4), (5, 5)])
+@pytest.mark.parametrize("padding", ("same", "valid"))
```
[the remark this answers is not in the document]
Comment (hunse) [reply nengo/builder/tests/test_transforms.py:116]: Again, I'd like to have the full test available for anyone making changes to this convolution code (covering so many different cases helped me to track down bugs, since there are so many potential edge cases in the code). But if you want to reduce things for the tests that typically run, I'm OK with that.

File: nengo/tests/test_transforms.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tbekolay) [reply nengo/tests/test_transforms.py:13]: Yeah it's tricky because they don't actually take a lot of time, it's just the sheer number of tests that it generates in the log. Maybe putting the tests in a separate file would be sufficient? Then it's a bit more obvious that this group of tests is doing something more specific that "testing transforms".

File: nengo/builder/tests/test_transforms.py [hunk 9]
```diff
@@ -109,9 +110,213 @@ def test_convinc_2d(
...
+
+                        assert success, f"Output {i} channel {k} differs"
+
+    assert success
+
+
+@pytest.mark.parametrize("x_size", (1, 2, 3, 4))
+@pytest.mark.parametrize("k_size", (1, 2, 3, 4))
+@pytest.mark.parametrize("stride", (1, 2, 3, 4))
+@pytest.mark.parametrize("padding", ("same", "valid"))
```
[the remark this answers is not in the document]
Comment (hunse) [reply nengo/builder/tests/test_transforms.py:259]: The other thing we can do if the concern is really noise in the test output, rather than the speed, is do the loop inside the test rather than with `parametrize`.

The reason I usually do [1, 2, 3, 4] is that 1 and 2 are fairly special (in that they're smaller than the typical values used for a lot of these parameters), and 3 and 4 give one even and one odd value in a more "normal" range (though you could also make an argument to use e.g. 6 and 9 instead of 3 and 4, but you'll still have 4 values for each one).

**Comment by hunse** [comment]
As for including TensorFlow, I'm happy to have that only run on release. I don't see a simple way to get around it; the whole point of vendorizing that `conv2d` library was that it's a pure Numpy implementation of convolution, so we don't need TensorFlow, but I then discovered that it wasn't always giving the same results as TensorFlow so I wanted to test against it. Now that I've fixed those bugs, it's probably not necessary for the tests to run every build, but it's still nice to have them if we're changing the code in the future (or if TensorFlow changes).

File: nengo/builder/tests/test_transforms.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tbekolay) [reply nengo/builder/tests/test_transforms.py:259]: Yeah that's fair, if they are useful cases then it makes sense to keep them. I think a different file would result in nicer output if there is a test failure, as it would tell you all the cases that fail rather than having to sort that out after the fact. Also, the simplest code would fail fast, and not failing fast results in a lot of bookkeeping code that is kind of what pytest is built to do. So I think moving it to a separate file is the way to go, at least for these 128 case ones.

**Comment by tbekolay** [comment]
Huh ... I'm not sure why exactly but this branch is causing the Appveyor build to [run out of memory](https://ci.appveyor.com/project/nengo/nengo/builds/41752122/job/qulgxxcv041u5kbr). Only for Python 3.6 though so ... not sure if it's worth debugging much.

File: nengo/builder/tests/test_transforms.py [hunk 10]
```diff
@@ -189,67 +189,68 @@ def test_convtransposeinc_2d(
...
 
-    # Check whether the outputs match. If they don't, print/plot some diagnostic output.
     success = allclose(conv_transpose_output, conv_output)
-    if not success:
-        # transpose convolution should act as the derivative for forward convolution,
-        # compute and compare to the derivative to see exactly where the difference is
-        signals = {sig: np.array(sig.initial_value) for sig in (x, w, y)}
-        step_conv = ConvInc(w, x, y, conv).make_step(signals, None, None)
-        step_conv()
+    # if not success:
```
Comment (hunse) [inline nengo/builder/tests/test_transforms.py:193]: Why comment this out?

Comment (tbekolay) [reply nengo/builder/tests/test_transforms.py:193]: Whoops that was debugging to make sure I was passing in the right things to the debug function. That might explain why it's running out of memory actually ...

File: nengo/builder/tests/test_transforms.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hunse) [reply nengo/builder/tests/test_transforms.py:259]: Yeah, a different file sounds good to me.

**Review by tbekolay — APPROVED** [review]
Moved the test, made a story for running with TF on release, and cleaned up the history. I'll merge on successful CI.
